### PR TITLE
Fuse delimiter support

### LINKED ISSUE #506: Hook new Delimiter API into fuse
We now support line / json / blob type delimiting when calling `PutFile()`

But to make this easily usable for our users, we need fuse to expose an interface to specify which delimiter you want to use. We've discussed the following:

- Use file suffixes to denote the delimiter type (.txt for line based / .json for json / .blob for no delimiting)
- By default, no suffix will imply line based delimiting
- To do this, we'll need fuse to check the suffix and update its calls for putfile
- Additionally, since FUSE makes several calls to `PutFile` at various times during normal file operations, we'll have to make sure to aggregate these putfiles server side into a single `PutFile` request to the block API
- I think we'll want to do this on a per-container basis not a per-job/commit basis
    - per-container - implies that each container's writes will constitute a single `PutFile()` request to the block API
        - e.g. in the no-delimiter case this means that each container will specify an indivisible blob of data per file
    - per commit - implies that all of the writes from all of the containers will be aggregated into a single `PutFile()` request to the block API
        - e.g. in the no-delimiter case this means that all of the writes to a single file for the whole commit will constitute a single blob of indivisible data

### PR DESCRIPTION
**Pull request by sjezewski**
Fixes #506 

- FUSE now looks at the file suffix to denote the delimiter it uses in the `PutFile` request
- json / none / line delimiters supported via `.json` / `.blob` / and `.txt` (or no suffix) respectively

We were concerned that the buffered writes from fuse -> PFS would cause trouble in particular for the `Delimiter_NONE` case.

I added a breaking test case explicitly for this, and it passes once FUSE detects suffix. Specifically, the reason that the buffered writes to PFS don't count as separate `PutFile()` requests, is a combination of two things:

1)  that we re-use the writer to the `PutFile()` request (see [here](https://github.com/pachyderm/pachyderm/blob/master/src/server/pfs/fuse/filesystem.go#L373)). This means that all write requests that fuse sees get aggregated together into the same `PutFile()` request

and

2) Fuse seems to call `Flush()` ([here](https://github.com/pachyderm/pachyderm/blob/master/src/server/pfs/fuse/filesystem.go#L409)) at the Right Times (in our case ... between when I write the file the first time and when I append to it) which means the putfile writer gets reset at the correct times

### CODE REVIEW

File: src/server/pfs/fuse/filesystem.go [hunk 1]
```diff
@@ -288,12 +288,22 @@ func (f *file) Fsync(ctx context.Context, req *fuse.FsyncRequest) error {
 	return nil
 }
 
+func (f *file) delimiter() pfsclient.Delimiter {
+	if strings.HasSuffix(f.File.Path, ".json") {
+		return pfsclient.Delimiter_JSON
+	}
+	if strings.HasSuffix(f.File.Path, ".blob") {
```
Comment (jdoliner) [inline src/server/pfs/fuse/filesystem.go]: Minor but I think `.bin` is a little more common than `.blob` for raw binary data. Could we do that instead.

**Comment by jdoliner** [comment]
All looks good save for the one comment (gif will be posted upon completion)

Did you try adding any tests that call `Flush` explicitly? I'm still pretty confident that we could get our delimiter code to break with that. But that was already the case before this change so it shouldn't prevent us merging this in.

File: src/server/pfs/fuse/filesystem.go (hunk 1 above)
Comment (sjezewski) [inline src/server/pfs/fuse/filesystem.go]: Yup! I think that's more descriptive - our suffixes describe the type of data we're storing.

**Comment by jdoliner** [comment]
tumblr_lqaynotksh1qggk3do1_500

**Comment by sjezewski** [comment]
Oh, and about the `Flush()` question -

I didn't write the flush test because while writing it, I got a failure (doing an `os.Open` and then `f.Write()` this was #521 - turns out I was just opening the file incorrectly.

Once that gets merged, we'll have that test.

**Pull request merged**